PT-FIXES: DOS-793: add locale-aware NumberParser and associated tests for number formatting and parsing

abdelaziz-mahdy opened the pull request:
updated version of https://github.com/kgrgreer/foam3/pull/4355

## Summary

Adds a grammar-based NumberParser following the same patterns as DateGrammar/DateParser, enabling locale-aware number parsing for CSV imports and FloatView.

- New `NumberGrammar` and `NumberParser` classes following FOAM's grammar/parser pattern
- Support for Standard (US/UK: `1,000.00`) and European (DE/FR: `1.000,00`) number formats
- Lenient parsing that accepts variable digit grouping (e.g., Indian-style `1,00,000`)
- Integration with Upload/Mapping system for CSV imports
- FloatView updated to use `foam.locale` for locale-aware formatting/parsing

## Number Format Support

### Standard Format (US/UK)
- Comma for thousands separator: `1,000,000`
- Period for decimal: `1,234.56`

### European Format (DE/FR)
- Period for thousands separator: `1.000.000`
- Comma for decimal: `1.234,56`

### Lenient Parsing
- Variable digit grouping: `1,00,000` (Indian style), `100,1000.1000`
- Any number of decimal places: `1000.123456789`
- Plain numbers without separators: `1234567.89`

## Usage

```javascript
var parser = foam.parse.NumberParser.create();

// Standard format (default)
parser.parseString('1,234.56');        // Returns: 1234.56

// European format
parser.parseString('1.234,56', 'european');  // Returns: 1234.56
```

abdelaziz-mahdy committed:
feat: add locale-aware NumberParser and associated tests for number formatting and parsing

jlhughes commented on src/foam/util/NumberParser.js:
```diff
@@ -0,0 +1,341 @@
+ */
+
+foam.CLASS({
+  package: 'foam.util',
```
Thinking, package foam/lib/parse (or foam/parse) is more suitable for additional parsers (and tests).

abdelaziz-mahdy replied:
done.

jlhughes commented on src/foam/util/NumberParser.js:
```diff
@@ -0,0 +1,341 @@
+    },
+    {
+      class: 'String',
+      name: 'initializedLocale_',
```
Why this naming format with the underscore?

abdelaziz-mahdy replied:
removed.

jlhughes commented on src/foam/util/NumberParser.js:
```diff
@@ -0,0 +1,341 @@
+
+        try {
+          // Ensure separators are initialized for current locale (lazy initialization)
+          if ( getInitializedLocale_() == null || !getInitializedLocale_().equals(getLocale()) ) {
```
```suggestion
          if ( getInitializedLocale_() == null || ! getInitializedLocale_().equals(getLocale()) ) {
```
and elsewhere.

abdelaziz-mahdy committed:
Update src/foam/util/NumberParser.js

Co-authored-by: Joel Hughes <jlhughes@users.noreply.github.com>

abdelaziz-mahdy committed:
feat: implement locale-aware NumberParser and associated tests

abdelaziz-mahdy commented:
Missing integration with Upload, will be added before merging, will be the same as date formats

kgrgreer commented:
fromCSV isn't used for JS, instead allow for alternate input mappings in Upload, just like for Dates
Comment-out unused java code.

abdelaziz-mahdy committed:
feat: enhance NumberParser with locale-aware formatting and parsing, add NumberFormat enum, and remove obsolete tests/javacode

abdelaziz-mahdy committed:
Merge with Dev

abdelaziz-mahdy committed:
feat: add NumberFormat views and parser for locale-aware number formatting

- Introduced NumberFormatCitationView and NumberFormatSelectionView for displaying number formats with descriptions.
- Added NumberFormatRichChoiceView for handling enum values related to number formats.
- Implemented NumberGrammar and NumberParser to support parsing and formatting of standard and European number formats.
- Created comprehensive tests for NumberParser covering various formats, including edge cases and invalid inputs.
- Removed legacy NumberParser implementation and associated tests.

abdelaziz-mahdy committed:
feat: refactor FloatView to use Intl.NumberFormat for locale-aware number formatting and remove obsolete locale property

abdelaziz-mahdy committed:
feat: add lenient parsing tests for variable digit grouping and extended decimals in NumberParser

abdelaziz-mahdy committed:
fix: update NumberParserTest flags to include java&test

abdelaziz-mahdy committed:
refactor: streamline fromCSV methods in Int and Float classes for better readability

abdelaziz-mahdy committed:
fix: update copyright year to 2026 in multiple files

abdelaziz-mahdy committed:
refactor: enhance fromCSV methods in Int and Float classes with improved parameter documentation and parsing logic
